### PR TITLE
refactor: add Ast.Constant and use in WeededAst

### LINKED ISSUE #4780: Introduce `Ast.Constant`
Introduce `Ast.Constant` as a new type:

```scala
    case class Unit(loc: SourceLocation) extends FinalAst.Expression {
      final val tpe = MonoType.Unit
    }

    case class Null(tpe: MonoType, loc: SourceLocation) extends FinalAst.Expression

    case class True(loc: SourceLocation) extends FinalAst.Expression {
      final val tpe = MonoType.Bool
    }

    case class False(loc: SourceLocation) extends FinalAst.Expression {
      final val tpe = MonoType.Bool
    }

    case class Char(lit: scala.Char, loc: SourceLocation) extends FinalAst.Expression {
      final val tpe = MonoType.Char
    }

    case class Float32(lit: scala.Float, loc: SourceLocation) extends FinalAst.Expression {
      final val tpe = MonoType.Float32
    }

    case class Float64(lit: scala.Double, loc: SourceLocation) extends FinalAst.Expression {
      final val tpe = MonoType.Float64
    }

    case class BigDecimal(lit: java.math.BigDecimal, loc: SourceLocation) extends FinalAst.Expression {
      final val tpe = MonoType.BigDecimal
    }

    case class Int8(lit: scala.Byte, loc: SourceLocation) extends FinalAst.Expression {
      final val tpe = MonoType.Int8
    }
/// ... rest
```

Then introduce e.g. `FinalAst.Cst` (just above Var).

### PR DESCRIPTION
**Pull request by mlutze**
fixes #4780

### CODE REVIEW

**Comment by mlutze** [comment]
@magnus-madsen I've only implemented this for the Weeder to ensure things look as they should.

File: main/src/ca/uwaterloo/flix/language/ast/Ast.scala [hunk 1]
```diff
@@ -79,6 +79,38 @@ object Ast {
...
   }
 
+  /**
+    * A common supertype for constant values.
+    */
+  sealed trait Constant
+  object Constant {
+    case object Unit extends Constant
+
+    case object Null extends Constant
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/ast/Ast.scala:89]: I assume the source locations are associated with the Expression?

File: main/src/ca/uwaterloo/flix/language/ast/WeededAst.scala [hunk 2]
```diff
@@ -87,33 +87,7 @@ object WeededAst {
...
-    case class Int16(lit: scala.Short, loc: SourceLocation) extends WeededAst.Expression
-
-    case class Int32(lit: scala.Int, loc: SourceLocation) extends WeededAst.Expression
-
-    case class Int64(lit: scala.Long, loc: SourceLocation) extends WeededAst.Expression
-
-    case class BigInt(lit: java.math.BigInteger, loc: SourceLocation) extends WeededAst.Expression
-
-    case class Str(lit: java.lang.String, loc: SourceLocation) extends WeededAst.Expression
+    case class Constant(cst: Ast.Constant, loc: SourceLocation) extends WeededAst.Expression
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/ast/WeededAst.scala:90]: Can we rename Constant to Cst? Then we have Cst and Var which is beautiful :)

Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/ast/WeededAst.scala:90]: Also it avoids a clash between the two.

File: main/src/ca/uwaterloo/flix/language/ast/Ast.scala (hunk 1 above)
Comment (mlutze) [reply main/src/ca/uwaterloo/flix/language/ast/Ast.scala:89]: Yes.

**Comment by magnus-madsen** [comment]
Wait... I want `Exp.Cst` and `Ast.Constant` or `Ast.Literal`.

**Comment by magnus-madsen** [comment]
:O

**Comment by magnus-madsen** [comment]
OK, so now we can start to remove code :)

**Pull request merged**